draft-ietf-ppm-dap.md

```diff
@@ -484,6 +480,133 @@ struct {
+
+The configuration of this query type specifies two paramters:
+
+- `min_batch_size` - Is the smallest number of measurements the batch is allowed
+  to include. In a sense, this parameter controls the degree of privacy that
+  will be obtained: The larger the minimum batch size, the higher degree of
+  privacy. However, this ultimately depends on the application and the nature of
+  the measurements and aggregation function.
+
+- `min_batch_duration` - The duration of the smallest permitted batch interval.
+  Additional restrictions apply; see {{batch-validation}}.
+
+The Collector is free to choose any sqeuence batch intervals it wishes, subject
+to the restrctions prescribed in {{batch-validation}}. However, it is
+RECOMMENDED that the sequence of batch intervals requested by the Collector all
+have the same duration, are continous, and increases monotonically. For example,
+the sequence of batch intervals `(1659544000, 1000)`, `(1659545000, 1000)`,
+`(1659545000, 1000)`, `(1659546000, 1000)` satisfies these conditions. (The
+first element of the pair denotes the start of the batch interval and the second
+denotes the duration.) There are execptions to this standard operating
+procedure:
+
+- The Collecotry may need to extend a batch interval forward in time if the
+  interval has elabsed and an insufficient number of measurements have been
+  uploaded. For example, instead of `(1659546000, 1000)`, the Collector might an
+  issue a query for `(1659546000, 2000)`.
+
+- Batch intervals may appear out-of-order if, for example, the Collector needs
+  to query a previous batch with a new aggregation parameter. This is permitted
+  for certain VDAFs.
+
+### Fixed-size Chunks
+
+The `fixed-chunk` query type is used to support applications in which the
+Collector needs the ability to strictly control the sample size. This is
+particularly important for controlling the amount of noise added to measurements
+by Clients (or added to aggregate shares by Aggregators) in order to achieve
+differential privacy.
+
+For this query type, the Aggregators batch measuremnet into arbitrary "chunks"
+such that each chunk has roughly the same size. The configuration includes
+parameters `min_batch_size` and `max_batch_size` that determine, respectively,
+the minumum and maximum number of measurements per chunk.
+
+To get the aggregate of the next chunk, the Colector issues a "chunk query" of
+type `next-chunk`. (See ChunkQuery in {{query}}.) The response to this query
+((see {{collect-flow}})) includes a "chunk ID", which can be used to query the
+same chunk later on. This is done using the `by-chunk-id` chunk query type.
+
+Implementation note: The goal for the Aggregators is to aggregate precisely
+`min_batch_size` measurements per chunk. Doing so, however, is challenging for
+Leader deployments in which multiple, independent nodes running the aggregate
+sub-protocol (see {{aggregate-flow}}) need to be coordinated. The maximum batch
```

Comment:
I think creating an aggregation job and attaching it to a given chunk requires knowing [an overestimate of] the sum of the number of reports successfully aggregated + the number of "outstanding" reports in in-progress aggregation jobs. Otherwise, creating an aggregation job may cause the chunk to grow past `max_batch_size`, effectively losing the chunk.

If you can update this sum "transactionally" with creation of the aggregation job, you can always hit `min_batch_size` exactly.

If you can't, I think avoiding chunk overflow requires having enough "leeway" (in `max_batch_size - min_batch_size`) such that every worker creating a max-size job at once will not move a chunk that has not yet aggregated `min_batch_size` reports to have aggregated more than `max_batch_size` reports, since every worker might simultaneously see the same information and independently decide to schedule an aggregation job. Another way of putting it is: I think increasing the "precision" of the size of your chunks would require reducing parallelism in the number of workers simultaneously aggregating into the chunk; equivalently, increasing worker parallelism to aggregate a task more quickly will lead to decreased "precision" in the size of the chunks that are generated.

Is that sort of what you're thinking (especially in the analysis in the previous paragraph)? I'm also curious if hitting `min_batch_size` exactly provides advantages over just knowing the number of reports in the chunk -- it would be relatively easy to communicate the report count to the collector, and if doing so would let us stop worrying about `max_batch_size` that might be the easier path.

Reply:
My bias is for simplicity in the protocol text, which I think argues for requiring fixed size chunks, and my intuition is that this should be achievable. My intuition (I'm using "intuition" as a weasel word here so that I don't actually have to support my arguments) is that the coordination you describe should be achievable in a datastore with [serializability](https://en.wikipedia.org/wiki/Serializability), which I think DAP implementations require anyway in order to do things like anti-replay.

However I think I want to get more implementation experience with the fixed chunk mode before I argue this point further, so let's go ahead with what's here now.

Finally, we already have #163 to wire up the report count to the collector, orthogonally to what's going on in this PR.

Reply:
@BranLwyd as discussed at IETF, the main reason to target a given batch size is that in some statistical applications you need to be able to control for the sample size. This would be important, for example, when using local differential privacy.

I'd like to start resolving threads where possible. The question left open here is whether requiring `min_batch_size == max_batch_size` is feasible. @tgeoghegan are you alright with resolving this with an OPEN ISSUE in the text?

Reply:
Yes, I'm comfortable revisiting this after we get some experience implementing the fixed chunk mode. However I would urge you to highlight this particular question in the thread you started on the ppm-ietf list just in case someone there has an opinion.

Reply:
Done.